docs/isl-2-0/spec.md

```diff
@@ -0,0 +1,673 @@
+
+### Ranges
+
+{% include grammar-element.md productions="range_int,range_number,range_timestamp,range_timestamp_precision,number,exclusivity" %}
+
+Some constraints can be defined by a range.
+A range is represented by a list annotated with `range`, and containing two values, in order: the minimum and maximum ends of the range.
+The default behavior is for both ends of the range to be *inclusive*; 
+if *exclusive* behavior is desired, the minimum or maximum (or both) values shall be annotated with `exclusive`.
+If the minimum or maximum end of a range is to be unspecified, this shall be represented by the symbols `min` or `max`, respectively; 
+the `exclusive`annotation is not applicable when the symbols `min` or `max` are specified.
+A range may not contain both `min` and `max`.
+
+All ranges have a type.
+The type of the range is the same as that of the minimum and/or maximum values specified in the range list. 
+If both a minimum and maximum values are specified (i.e. `min` and `max` are not used), then both of those values must be of the same type. 
+(For example, `range::[1995-12-06T, 55.4]` mixes values of the timestamp and number types, and therefore is not a valid range.)
+
+
+## all_of
+
+{% include grammar-element.md productions="all_of" %}
+
+Value must be valid for all the types listed. The list of types must not be empty.
+
+{% include example.md title="`all_of` constraint" markdown="
+```ion
+type::{
+  name: Duck,
+  all_of: [
+    WalksLikeADuck,
+    QuacksLikeADuck,
+  ]
+}
+```
+" %}
+
+## annotations
+
+{% include grammar-element.md productions="annotations,annotations_modifier" %}
+
+Restricts the allowed annotations on a value.
+Applicable to all types except `document` (since documents cannot be annotated).
+This constraint has two available syntaxes.
+The "Standard" syntax is very expressive, but can be verbose for simple use cases.
+The "Simple" syntax is based on the Ion Schema 1.0 `annotations` syntax, and is concise, but only supports a limited set of use cases.
+
+### Standard Syntax
+
+The `annotations` constraint specifies the type and/or constraints for all annotations of a value.
+[The annotations themselves are symbols _tokens_](https://amzn.github.io/ion-docs/docs/spec.html#annot) but will be represented as a non-null list of non-null, un-annotated symbol _values_ for the purpose of validation.
+The list of annotation symbols must match the given type.
+
+{% capture sample_code %}
+annotations: { contains: [red, blue] }                   // Annotations must contain "red" and "blue" in any order, and may contain other annotations, such as "yellow"
+annotations: { element: { valid_values: [red, blue] } }  // Only the annotations "red" and "blue" are permitted, but they are not required
+annotations: { container_length: 0 }                     // No annotations are permitted
+annotations: { element: { regex: "\\w+(\\.\\w+)" } }     // Annotations must match the regex
```

Comment:
I seem to recall that using the `ordered_elements` constraint on `annotations` was one of the motivations for the expanded syntax in ISL 2.0. Is that correct? If so, it would be good to include an example of that.

Reply:
The ISL 1.0 syntax already had an `ordered` annotation for the `annotations` constraint, but I think the expanded syntax helps bring some consistency to how we handle ordered things.

The main motivation was that the ISL 1.0 syntax doesn't allow you to specify any constraint-based rule for the annotation _text_—you could only specify a literal annotation text—nor any rule for the number of annotations.

I'll add a few more examples.